### PR TITLE
Reimplement verifyAllCombinations as variadic function

### LINKED ISSUE #33: Increase the number of collections in `verifyAllCombinations` beyond 9
Currently, the number of collections in `verifyAllCombinations` is limited to 9, and is extensible by code by the user of the library.
Some legacy code has functions with more than 9 parameters, sometimes many more. It would be more convenient if the user didn't have to write the extension themselves.
This issue suggests to increase the number of collections that the library can support natively, ideally without limit, by using variadic templates.
If I'm not mistaken, Clare already mentioned this could be an evolution for the library. This issue is then to log this request.
Keep up the amazing work! :)

### PR DESCRIPTION
**Pull request by mika-fischer**
Fixes #33

The parameter order needed to change, since parameter pack types
can only be deduced if the pack is the last parameter.

For backwards compatibility overloads with the reporter at the end
are provided for up to 9 inputs (as was the case before).

Also put the functions in a namespace instead of a class.

### CODE REVIEW

File: tests/Catch2_Tests/CombinationTests.cpp [hunk 1]
```diff
@@ -67,3 +67,64 @@ TEST_CASE("YouCanVerifyCombinationsOf9") {
...
+        );
+    }
+    catch(const ApprovalException&)
+    {
+        // ignore
+    }
+    REQUIRE(reporter.called == true);
+}
+
+TEST_CASE("YouCanVerifyCombinationsOf9ReportsOldParameterOrder")
```
Comment (claremacrae) [inline tests/Catch2_Tests/CombinationTests.cpp]: There's something I'm missing here...
There's a new test with 'OldParameterOrder' in the name, but the existing tests have not been changed - which suggests to me that they are also using the old order still.

File: ApprovalTests/CombinationApprovals.h [hunk 2]
```diff
@@ -2,363 +2,260 @@
...
+void verifyAllCombinations(Converter&& converter, const Reporter& reporter, const Container& input0, const Containers&... inputs)
+{
+    std::stringstream s;
+    Detail::cartesian_product(Detail::serialize<Converter>{s, std::forward<Converter>(converter)}, input0, inputs...);
+    Approvals::verify(s.str(), reporter);
+}
+
+template<class Converter, class... Containers>
+Detail::enable_if_t<!Detail::disjunction<std::is_base_of<Reporter, Containers>...>::value>
+verifyAllCombinations(Converter&& converter, const Containers&... inputs)
```
[the remark this answers is not in the document]
Comment (claremacrae) [reply ApprovalTests/CombinationApprovals.h]: It just works - via magic that I don't yet understand!

I've found that if I pass in a custom reporter before the contains of values, it just works - amazing.

**Comment by mika-fischer** [comment]
Hi Clare,

thanks for taking a look!

regarding the reporter parameter: There are now two main overloads:
- First, the one with a reporter parameter at the front, which is the one that does the work. It only matches if the second argument is a reporter:
```cpp
template<class Converter, class Container, class... Containers>
void verifyAllCombinations(Converter&& converter, const Reporter& reporter, const Container& input0, const Containers&... inputs);
```
- Then, there's this one which only matches if there's no reporter argument anywhere (i.e. neither before the inputs nor after the inputs) and uses the DefaultReporter. That means this overload replaces the reporter default arguments from before:
```cpp
template<class Converter, class... Containers>
Detail::enable_if_t<!Detail::disjunction<std::is_base_of<Reporter, Containers>...>::value>
verifyAllCombinations(Converter&& converter, const Containers&... inputs);
```

And then we have the backwards compatibility overloads (up to 9 inputs), which only match if the last argument is a reporter. They don't need a default argument for the reporter because, if no reporter is given, the overload above already matches:
```cpp
template<class Converter, class C1>
void verifyAllCombinations(Converter&& converter, const C1& i1, const Reporter& reporter);

template<class Converter, class C1, class C2>
void verifyAllCombinations(Converter&& converter, const C1& i1, const C2& i2, const Reporter& reporter);
...
```

But because we can't have parameters after a parameter pack we cannot use a variadic function and are back to repetitious code, which is not so nice.

So my thought was to deprecate the old parameter order with the reporter at the end in favor of the new order with it before the inputs. And then in the next major release the backwards comatibility overloads could just be removed.

Also, if C++11 support should ever be dropped, this could be made a lot simpler.

Open questions:
- Maybe the reporter parameter should go even before the converter parameter?

File: ApprovalTests/CombinationApprovals.h [hunk 3]
```diff
@@ -2,363 +2,260 @@
...
+{
+    verifyAllCombinations(std::forward<Converter>(converter), DefaultReporter(), inputs...);
+}
+
+// These are here for backwards compatibility
+// Maybe they should be deprecated and removed in the future?
+template<class Converter, class C1>
+void verifyAllCombinations(Converter&& converter, const C1& i1, const Reporter& reporter)
```
[the remark this answers is not in the document]
Comment (mika-fischer) [reply ApprovalTests/CombinationApprovals.h]: See below, this is handled via function overlods now

File: ApprovalTests/CombinationApprovals.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (mika-fischer) [reply ApprovalTests/CombinationApprovals.h]: See below for an explanation

File: tests/Catch2_Tests/CombinationTests.cpp (hunk 1 above)
Comment (mika-fischer) [reply tests/Catch2_Tests/CombinationTests.cpp]: Yes, you're right, thanks!

However, there's actually just one test case that uses a custom reporter. For the others there really no change.

I can also duplicate that test to check both the old and new forms.

But let's first discuss where the reporter parameter should actually be. Before or after the converter?

**Comment by claremacrae** [comment]
Hi, thanks again for this - and sorry we haven't got back to you yet... I've only just got back from travels after CppCon, and haven't yet been able to speak with @isidore to talk through this, and think about the parameter order...

I have enjoyed reading through the code more carefully, and have a good idea of how it works - I learned a lot, thank you.

All your comments make sense to me, too...

Thanks again!

**Comment by mika-fischer** [comment]
> I have enjoyed reading through the code more carefully, and have a good idea of how it works - I learned a lot, thank you.

Great, glad to help! Let me know if you need anything else.

**Comment by claremacrae** [comment]
Code review request:

@mika-fischer If you have time, please could you have a look at this branch on my fork of ApprovalTests.cpp? It contains edits I made whilst working out how the code in this PR worked...

https://github.com/claremacrae/ApprovalTests.cpp/tree/pr34_experiments

A summary is:
* I moved all the Cartesian product code out to a separate `ApprovalTests/utilities/CartesianProduct.h` header - so I could see what stuff was for Cartesian product, and what was for CombinationApprovals - and with hindsight, to separate concerns
* I added some explanatory comments to the above, whilst working through it (I don't understand why `dereference_iterator` is needed)
* I wrote some tests of cartesian_product, to confirm my understanding about its behaviour

The "Files changed" tab in this comparison may be a good way to view the differences:
https://github.com/claremacrae/ApprovalTests.cpp/compare/pr/34...pr34_experiments

I would be grateful for any comments you have about this branch. 

In particular, how comfortable would you be with it being adopted, after this pull request is accepted?

(For example, I don't know if it would create conflicts against any later changes you might have made in your own code...)

**Comment by mika-fischer** [comment]
Hi @claremacrae,

This all looks fine to me. There's just one comment that's a bit misleading:
```cpp
// Move on to next row, for when only one input container was supplied
template<class Its, std::size_t I = tuple_size<Its>()-1>
enable_if_t<I == 0>
increment_iterator(Its& it, const Its&, const Its&) {
    ++std::get<I>(it);
}

// Move on to next row, for when more than one input container was supplied
template<class Its, std::size_t I = tuple_size<Its>()-1>
enable_if_t<I != 0>
increment_iterator(Its& its, const Its& begins, const Its& ends) {
    if (++std::get<I>(its) == std::get<I>(ends)) {
        std::get<I>(its) = std::get<I>(begins);
        increment_iterator<Its, I-1>(its, begins, ends);
    }
}
```

In fact, the first overload is needed in all cases. It's used to terminate the recursion present in `increment_iterator`, which works by incrementing the last iterator, and if that wraps around, then incrementing the second-to-last iterator and so on, and that recursion needs to be terminated at the first iterator.

So better comments would be:
- Increment outermost iterator. If it reaches end, we're finished and do nothing.
- Increment inner iterator. If it reaches its end, we reset it and increment the previous iterator.

Regarding the `dereference_iterator` struct, it's needed to transform the iterators into value references which are then passed to the visitor function:
```cpp
Detail::apply(std::forward<F>(f), Detail::transform<Detail::dereference_iterator>(its));
```

Does that answer your question? 

Other than that, you can just go ahead as you see fit. I have no changes in my branch and also don't plan any.

**Comment by claremacrae** [comment]
Awesome, thanks @mika-fischer - I've put those explanations in 6397bf3c5e3ffe138e6d7e72e1eda11c9256bec9, with attribution.

I really like your mention of Visitor - that's a good description of the F template parameter to `cartesian_product()` and related...

**Pull request merged**